File: valid.pl
#!/usr/bin/perl
use v5.24;

sub valid_group {
    my ($group_string) = @_;
    my @elems = split /\s+/, $group_string;
    my %elems = map { $_, 1 } @elems;
    return keys %elems == 1;
}

while (<>) {
    if ($_ =~ /^([a-zA-Z_\d]+) # test name 
	\s+(\d+) # test number 
	[\s\d]+
	(\(.*\)) #paren groups 
	$/x) {
	my $t_name = $1;
	my $t_num = $2;
	my @paren_groups = $3 =~ /\((.*?)\)/g;
	foreach my $group (@paren_groups) {
	    if (valid_group $group) {
		print "$t_name ($t_num) pass\n";
	    } else {
		print "$t_name ($t_num) FAIL\n";
	    }
	}
    } 
}

File: valid.py
#!/usr/local/bin/python3.6

import itertools
import sys


def groups_from_result_string(res_str):
    try:
        left = res_str.index('(')
        right = res_str.rindex(')')
        groups_str = res_str[left+1:right]
        groups = groups_str.split(') (')
        return groups
    except ValueError:
        return []


def name_from_result_string(res_str):
    return res_str.split()[0]


def group_passes(group):
    def all_zs(s):
        return s[0] == 'z' and len(set(s)) == 1
    elems = group.split()
    if any(all_zs(elem) for elem in elems):
        return True
    return len(set(elems)) == 1


class Result(object):
    def __init__(self, result_string):
        self.groups = groups_from_result_string(result_string)
        self.name = name_from_result_string(result_string)


    def analyze(self):
        for group in self.groups:
            if group_passes(group):
                print('{} pass'.format(self.name))
            else:
                print('{} FAIL'.format(self.name))


if __name__ == '__main__':
    for result_string in map(str.rstrip, sys.stdin):
        res = Result(result_string)
        res.analyze()
